src/EPub.ts

```diff
@@ -0,0 +1,244 @@
+import Store from "./Store";
+
+type xmlObject = {
+  [key: string]: string[] | string | xmlObject | [];
+};
+
+function xmlToJson(xml: any) {
+  let obj = {} as xmlObject;
+
+  // process ELEMENT_NODE
+  if (xml.nodeType == 1) {
+    if (xml.attributes.length > 0) {
+      obj["@attributes"] = {};
+      for (let j = 0; j < xml.attributes.length; j++) {
+        const attribute = xml.attributes.item(j);
+        obj["@attributes"][attribute.nodeName] = attribute.nodeValue;
+      }
+    }
+  }
+  // process TEXT_NODE
+  else if (xml.nodeType == 3) {
+    obj = xml.nodeValue;
+  }
+
+  if (xml.hasChildNodes()) {
+    for (let i = 0; i < xml.childNodes.length; i++) {
+      const item = xml.childNodes.item(i);
+      const nodeName: string = item.nodeName;
+
+      if (typeof obj[nodeName] == "undefined") {
+        obj[nodeName] = xmlToJson(item);
+      } else {
+        //@ts-ignore
+        if (typeof obj[nodeName].push == "undefined") {
+          const old = obj[nodeName];
+          obj[nodeName] = [];
+          //@ts-ignore
+          obj[nodeName].push(old);
+        }
+        //@ts-ignore
+        obj[nodeName].push(xmlToJson(item));
+      }
+    }
+  }
+  return obj;
+}
+export interface Metadata {
+  title?: string;
+  author?: string;
+  identifier?: string;
+  language?: string;
+  modified?: string;
+}
+
+export interface Link {
+  rel?: Array<string>;
+  href?: string;
+  type?: string;
+  title?: string;
+  children?: Array<Link>;
+}
+
+export default class Manifest {
+  public readonly metadata: Metadata;
+  public readonly links: Array<Link>;
+  public readonly spine: Array<Link>;
+  public readonly resources: Array<Link>;
+  public readonly toc: Array<Link>;
+  private readonly manifestUrl: URL;
+
+  public static async getManifest(
+    manifestUrl: URL,
+    store?: Store
+  ): Promise<Manifest> {
+    const fetchManifest = async (): Promise<Manifest> => {
+      const manifestJSON = await window
+        .fetch(manifestUrl.href)
+        .then((response) => response.text())
+        .then((str) => new window.DOMParser().parseFromString(str, "text/xml"))
+        .then((data) => JSON.stringify(xmlToJson(data)));
+
+      if (store) {
+        await store.set("manifest", JSON.stringify(manifestJSON));
+      }
+
+      return new Manifest(JSON.parse(manifestJSON), manifestUrl);
+    };
+
+    const tryToUpdateManifestButIgnoreResult = async (): Promise<void> => {
```

Comment:
Some of these methods are exactly the same as `Manifest.ts`
I wonder if there's a way to either rework `Manifest.ts` to recognize if it's an epub or not, or extract similar functions, or find some other way to consolidate this code so we don't have two copies of very similar code.  It'll also make it more obvious which parts of this are specific to the webpub stuff and which are specific to epub.

Reply:
I updated to have all of the code in one file for parsing manifest.json and package.opf in the same file. I wonder if it'll still be confusing to refer to both formats as Manifest

Reply:
yeah definitely, it's not really "manifest" as it is some kind of ... parser for the manifest-type file?  Aaaah names are hard.

Reply:
I renamed the parsed `package.opf` to `OPFPackage` in certain places for now.